### DISCUSSION TITLE
Apps & Features vs. winget list

### DISCUSSION THREAD

**Discussion by marco-eckstein** [microsoft/winget-pkgs#132064] [Q&A]
I am new to WinGet and have just submitted two packages: The new package [Line6.Line6Monkey](https://github.com/microsoft/winget-pkgs/tree/master/manifests/l/Line6/Line6Monkey/1.77) and an update to package [ArobasMusic.GuitarPro.7](https://github.com/microsoft/winget-pkgs/tree/master/manifests/a/ArobasMusic/GuitarPro/7/7.6.0.2089).

Both packages have non-trivial entries in _Apps & Feature_, and I am puzzled about the relationship between these and the packages. 

**Line6.Line6Monkey**

Manifest snippets:

```yaml
PackageIdentifier: Line6.Line6Monkey
PackageVersion: "1.77"
Publisher: Line 6, Inc.
PackageName: Line 6 Monkey
```

```yaml
AppsAndFeaturesEntries:
- DisplayName: Line 6 Uninstaller
  Publisher: Line 6
```

Apps & Features:

```
Name                  Publisher    Version
Line 6 Uninstaller    Line 6       <empty>
```

`winget list` output:

```
Name                  ID                   Version    Available   Source
Line 6 Uninstaller    Line6.Line6Monkey    Unknown    1.77        winget
```

**ArobasMusic.GuitarPro.7**

Manifest snippets:

```yaml
PackageIdentifier: ArobasMusic.GuitarPro.7
PackageVersion: 7.6.0.2089
Publisher: Arobas Music
PackageName: Guitar Pro 7
```

```yaml
AppsAndFeaturesEntries:
- DisplayName: Guitar Pro 7
- DisplayName: Guitar Pro 7 - Soundbanks
  DisplayVersion: 1.1.123
```

Apps & Features:

```
Name                         Publisher       Version
Guitar Pro 7                 Arobas Music    7.6.0.2089
Guitar Pro 7 - Soundbanks    Arobas Music    1.1.123
```

`winget list` outputs:

```
Name                         ID                         Version         Available     Source
Guitar Pro 7                 ArobasMusic.GuitarPro.7    > 7.6.0.2089                  winget
Guitar Pro 7 - Soundbanks    ArobasMusic.GuitarPro.7    7.6.0.2089                    winget
```

**Questions**

- Apparently, the name column in `winget list` is always the Apps & Features name and not the `PackageName`. Correct?
  - Wouldn't it make sense to add another (optional) column with the `PackageName`?
- Apparently, the version column in `winget list` is the Apps & Features version, unless `DisplayVersion` maps it to`PackageVersion`. Correct?
  - Sometimes seeing the version from App & Features and sometimes the `PackageVersion` can be confusing. Wouldn't it make sense to add two (optional) columns that show either?
- For `ArobasMusic.GuitarPro.7` | `Guitar Pro 7`, `PackageVersion` and the version in App & Features are the same, namely `7.6.0.2089`. So why does WinGet show version `> 7.6.0.2089`?
- Is there any way to let winget know that a certain version is installed, even though the version in Apps & Features is empty?
- To uninstall  `ArobasMusic.GuitarPro.7`, I have to run `winget uninstall --id  ArobasMusic.GuitarPro.7` twice. I do not think this is desired behavior. So is this a bug in my manifest or in WinGet?

It seems that the relationship between Apps & Features and packages can get quite complicated. Extensive docs would be desirable, but I could not find much.

**Comment by Trenly** [comment]
> Apparently, the name column in winget list is always the Apps & Features name and not the PackageName. Correct?

`winget list` is (basically) a dump of the data in Apps & Features. The Name here is the one from Apps & Features since not all packages correlate to a manifest

> Wouldn't it make sense to add another (optional) column with the PackageName?

Maybe, but it could cause more confusion than just showing what is in Apps & Features

> Apparently, the version column in winget list is the Apps & Features version, unless DisplayVersion maps it toPackageVersion. Correct?

Correct

> Sometimes seeing the version from App & Features and sometimes the PackageVersion can be confusing. Wouldn't it make sense to add two (optional) columns that show either?

It can be confusing either way. If a package is marketed as version `1.2.3` but writes `100.22.305.1` to ARP entries, then using Apps & Features mapping to the Package version is better representative of which version is installed when comparing to what is on a publishers website. Or, when the ARP version is a git commit hash, using the Package Version is easier to understand which is newer and which is older.

> For ArobasMusic.GuitarPro.7 | Guitar Pro 7, PackageVersion and the version in App & Features are the same, namely 7.6.0.2089. So why does WinGet show version > 7.6.0.2089?

Because there is probably an older manifest in the repository that specified Apps & Features entries and someone didn’t put them in the latest manifest. Adding the `AppsAndFeaturesEntries` to the manifests that are missing them should fix the issue. 
* See this bug for details - https://github.com/microsoft/winget-cli/issues/2503

Either that, or you have a newer version of the package installed also. Without seing exact output and verbose logs of `winget list "Guitar Pro"` from your specific setup, it's impossible to know the root cause.

> Is there any way to let winget know that a certain version is installed, even though the version in Apps & Features is empty?

No. If the version is empty, then WinGet has no way to know what is installed. The only way to tell WinGet the version would be to manually change what is written in the Apps & Features entry

> To uninstall  ArobasMusic.GuitarPro.7, I have to run winget uninstall --id  ArobasMusic.GuitarPro.7 twice. I do not think this is desired behavior. So is this a bug in my manifest or in WinGet?

It's possible you have multiple versions of GuitarPro installed. It's also possible that two apps with similar Apps & Features entries could be causing issues. Without knowing exactly what you have installed, having verbose logs, and copies of the manifests, it's impossible to know the exact cause. 

Edit: Looking at your manifest snippet more closely, its because you have two components for Guitar Pro 7. Each entry in Apps & Features is *technically* a separate program, and should (ideally) have its own PackageIdentifier and manifest. You have to call uninstall twice because `winget uninstall` can only call the uninstaller for one program. Also, when there are multiple items listed in `AppsAndFeaturesEntries` in the manifest, it can cause matching issues - only the primary Apps & Features entry should be added for now. 

If an installer installs multiple components to Apps & Features, then the publisher should update their uninstaller to also uninstall all the components

**Reply by marco-eckstein** [reply]
> > For ArobasMusic.GuitarPro.7 | Guitar Pro 7, PackageVersion and the version in App & Features are the same, namely 7.6.0.2089. So why does WinGet show version > 7.6.0.2089?
>
> Because there is probably an older manifest in the repository that specified Apps & Features entries and someone didn’t put them in the latest manifest. Adding the AppsAndFeaturesEntries to the manifests that are missing them should fix the issue.

No, there is no other manifest for this package.

> Either that, or you have a newer version of the package installed also. Without seing exact output and verbose logs of winget list "Guitar Pro" from your specific setup, it's impossible to know the root cause.

No - Apps & Features shows 7.6.0.2089.

Maybe it is because I did not state the `DisplayVersion` in `AppsAndFeaturesEntries`, thinking the `PackageVersion` would be inherited as `DisplayVersion`?

```yaml
AppsAndFeaturesEntries:
- DisplayName: Guitar Pro 7
- DisplayName: Guitar Pro 7 - Soundbanks
  DisplayVersion: 1.1.123
```

> Edit: Looking at your manifest snippet more closely, its because you have two components for Guitar Pro 7. Each entry in Apps & Features is technically a separate program, and should (ideally) have its own PackageIdentifier and manifest. You have to call uninstall twice because winget uninstall can only call the uninstaller for one program. Also, when there are multiple items listed in AppsAndFeaturesEntries in the manifest, it can cause matching issues - only the primary Apps & Features entry should be added for now.

OK, I've separated them. See https://github.com/microsoft/winget-pkgs/pull/132557 and https://github.com/microsoft/winget-pkgs/pull/132558. Since `AppsAndFeaturesEntries` is not needed anymore, the version issue above should also be solved.
